File: dawn_go/hash_5wi.go
import (
    "math/rand"
)

func universalHash(a, b, c, d, e, key int) int {
    m := 1 << 32
    hash := a*key + b*key*key + c*key*key*key + d*key*key*key*key + e*key*key*key*key*key
    return hash % m
}

func linearProbe(table []interface{}, key int) int {
    m := len(table)
    a, b, c, d, e := rand.Intn(m), rand.Intn(m), rand.Intn(m), rand.Intn(m), rand.Intn(m)
    hash := universalHash(a, b, c, d, e, key)
    for i := 0; ; i++ {
        index := (hash + i) % m
        if table[index] == nil {
            return index
        }
    }
}

File: dawn_rust/src/hash_5wi.rs
use rand::Rng;

fn universal_hash(a: u32, b: u32, c: u32, d: u32, e: u32, key: u32) -> u32 {
    let m = 1 << 32;
    let hash = (a * key
        + b * key.overflowing_mul(key).0
        + c * key.overflowing_mul(key.overflowing_mul(key).0).0
        + d * key.overflowing_mul(key.overflowing_mul(key.overflowing_mul(key).0).0).0
        + e * key.overflowing_mul(key.overflowing_mul(key.overflowing_mul(key.overflowing_mul(key).0).0).0).0)
        % m;
    hash
}

fn linear_probe<T>(table: &mut [Option<T>], key: u32) -> usize {
    let m = table.len();
    let mut rng = rand::thread_rng();
    let a = rng.gen_range(0..m as u32);
    let b = rng.gen_range(0..m as u32);
    let c = rng.gen_range(0..m as u32);
    let d = rng.gen_range(0..m as u32);
    let e = rng.gen_range(0..m as u32);
    let hash = universal_hash(a, b, c, d, e, key) as usize;
    for i in 0.. {
        let index = (hash + i) % m;
        if table[index].is_none() {
            return index;
        }
    }
    unreachable!()
}